site/content/posts/2023-08-09-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,337 @@
+For more information on this, see the [cert-manager docs](https://cert-manager.io/docs/configuration/selfsigned/#bootstrapping-ca-issuers).
+The `Certificate` with name `impersonation-serving-cert` will generate the leaf certificate used by the impersonation proxy to serve TLS.
+
+```shell
+$ cat << EOF > self-signed-cert.yaml
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: cert-manager
+---
+apiVersion: cert-manager.io/v1
+kind: ClusterIssuer
+metadata:
+  name: selfsigned-cluster-issuer
+spec:
+  selfSigned: {}
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: my-selfsigned-ca
+  namespace: cert-manager
+spec:
+  isCA: true
+  commonName: my-selfsigned-ca
+  secretName: self-signed-ca-for-kind-testing
+  privateKey:
+    algorithm: ECDSA
+    size: 256
+  issuerRef:
+    name: selfsigned-cluster-issuer
+    kind: ClusterIssuer
+    group: cert-manager.io
+---
+apiVersion: cert-manager.io/v1
+kind: ClusterIssuer
+metadata:
+  name: my-ca-issuer
+spec:
+  ca:
+    secretName: self-signed-ca-for-kind-testing
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: impersonation-serving-cert
+  namespace: pinniped-concierge
+spec:
+  secretName: impersonation-proxy-tls-serving-cert
+
+  duration: 2160h # 90d
+  renewBefore: 360h # 15d
+  subject:
+    organizations:
+    - Pinniped
+  isCA: false
+  privateKey:
+    algorithm: RSA
+    encoding: PKCS1
+    size: 2048
+  usages:
+  - server auth
+  dnsNames:
+  - impersonation-proxy-mtls.local
+  issuerRef:
+    name: my-ca-issuer
+    kind: ClusterIssuer
+    group: cert-manager.io
+
+EOF
+
+$ kubectl apply \
+    --filename self-signed-cert.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Download the root (self-signed) CA's certificate so that it can be advertised as the CA bundle for the Concierge impersonation proxy:
+
+```shell
+$ kubectl get secret self-signed-ca-for-kind-testing \
+    --namespace cert-manager \
+    --output jsonpath="{.data.ca\.crt}" \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    | tee self-signed-ca-for-kind-testing.pem.b64
+    
+# Tip: Put the contents of self-signed-ca-for-kind-testing.pem.b64 into your copy buffer for a later step!
+```
+
+Now update the `CredentialIssuer` to use the impersonation proxy (which is disabled on kind by default):
+
+```shell
+$ kubectl edit credentialissuer pinniped-concierge-config \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Make sure that the spec has the following values:
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: # paste the contents of the file self-signed-ca-for-kind-testing.pem.b64
+        secretName: impersonation-proxy-tls-serving-cert
+...
+# Now save and close the text editor
+
+# Confirm that the CredentialIssuer looks as expected
```

Comment:
Again note that the `status.strategies[.frontend.my-ca-issuer.my-ca-issuer]` should advertise your CA certificate (from the previous `kubectl get secret self-signed-ca-for-kind-testing`) and not the original self-signed CA certificate.

Reply:
Resolved locally.